presto-spi/src/main/java/com/facebook/presto/spi/type/TypeSignature.java

```diff
@@ -357,13 +358,31 @@ public boolean equals(Object o)
 
         TypeSignature other = (TypeSignature) o;
 
+        // TODO remove this hack together with hack from toString()
+        if (magicVarcharEquals(other, this)) {
```

Comment:
Why do we need such hack in `TypeSignature`? Can't we hack it somewhere when it is actually needed instead of changing the `TypeSignature` equals?